File: schemas/analytics-metrics/convex/dashboard_widgets.ts
// dashboard_widgets: Widgets within dashboards.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const dashboard_widgets = defineTable({
  dashboardId: v.id("dashboards"),
  metricId: v.optional(v.id("metric_definitions")),
  title: v.optional(v.string()),
  chartType: v.union(
    v.literal("line"),
    v.literal("bar"),
    v.literal("area"),
    v.literal("pie"),
    v.literal("number"),
    v.literal("table"),
    v.literal("funnel"),
    v.literal("map"),
  ),
  config: v.optional(v.any()),
  positionX: v.number(),
  positionY: v.number(),
  width: v.number(),
  height: v.number(),
  updatedAt: v.number(),
})
  .index("by_dashboard_id", ["dashboardId"])
  .index("by_metric_id", ["metricId"]);

File: schemas/analytics-metrics/spacetimedb/dashboard_widgets.rs
// dashboard_widgets: Individual widgets placed on dashboards with chart type and position.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum ChartType {
    Line,   // type: String
    Bar,
    Area,
    Pie,
    Number,
    Table,
    Funnel,
    Map,
}

#[spacetimedb::table(name = dashboard_widgets, public)]
pub struct DashboardWidget {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub dashboard_id: String, // UUID, FK → dashboards.id (cascade delete)

    #[index(btree)]
    pub metric_id: Option<String>, // UUID, FK → metric_definitions.id (set null)

    pub title: Option<String>,
    pub chart_type: ChartType,
    pub config: Option<String>, // JSON
    pub position_x: i32,
    pub position_y: i32,
    pub width: i32,
    pub height: i32,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
